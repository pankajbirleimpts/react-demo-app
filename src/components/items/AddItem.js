import React, { Component } from 'react';
import { withRouter, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import {
  Formik, Form, Field, ErrorMessage,
} from 'formik';
import * as Yup from 'yup';
import { Grid } from 'semantic-ui-react';
import { getItem, updateItem, addItem } from '../../actions/ItemAction';
import { langs, itemCategories } from '../../config';
import { Loader } from '../common';


export class UnconnectedAddItem extends Component {
  state = {
    id: '',
    initialValues: {
      itemName: '',
      description: '',
      category: '',
      amount: '',
      isActive: '',
    },
  }

  componentDidMount() {
    const { id } = this.props.match.params;
    if (id) {
      this.props.getItem(id, (response) => {
        this.setState({
          initialValues: response,
          id,
        });
      });
    }
  }

  /**
   * @formSubmitHandler
   * @desc: Submit the signup form
   */
  formSubmitHandler = (values) => {
    const { id } = this.props.match.params;
    if (id) {
      // Update item
      this.props.updateItem(id, values, () => {
        this.props.history.replace('/items');
      }, true);
    } else {
      // Add Item
      this.props.addItem(values, () => {
        this.props.history.replace('/items');
      });
    }
  };

  /**
   * @formValidation
   * @desc: form validation
   */
  formValidation = () => Yup.object().shape({
    itemName: Yup.string()
      .max(25, langs.messages.CHAR_MAX_LIMIT_25)
      .required(langs.messages.REQUIRED),
    description: Yup.string()
      .max(300, langs.messages.CHAR_MAX_LIMIT_300)
      .required(langs.messages.REQUIRED),
    category: Yup.string()
      .required(langs.messages.REQUIRED),
    amount: Yup.number()
      .required(langs.messages.REQUIRED),
    isActive: Yup.string()
      .required(langs.messages.REQUIRED),
  });

  /**
   * @renderForm
   * @desc: rednder form
   */
  renderForm = ({ values, setFieldValue }) => {
    const itemOptions = itemCategories.map(val => <option value={val.value}>{val.label}</option>);
    return (
      <Form noValidate className="ui form" >
        <div className="field">
          <label>Item Name *</label>
          <Field data-test="itemName-input" name="itemName" type="text" />
          <ErrorMessage component="p" name="itemName" className="red" />
        </div>
        <div className="field">
          <label>Category *</label>
          <Field data-test="category-input" name="category" as="select">
            <option value="">Please Select</option>
            {itemOptions}
          </Field>
          <ErrorMessage component="p" name="category" className="red" />
        </div>
        <div className="field">
          <label>Amount *</label>
          <Field data-test="amount-input" name="amount" type="number" />
          <ErrorMessage component="p" name="amount" className="red" />
        </div>
        <div className="field">
          <label>Description *</label>
          <Field data-test="description-input" name="description" as="textarea" />
          <ErrorMessage component="p" name="description" className="red" />
        </div>
        <div className="field">
          <label>Status *</label>
          <Field data-test="isActive-input" name="isActive" as="select">
            <option value="">Please Select</option>
            <option value="Active">Active</option>
            <option value="Deactivev">Deactive</option>
          </Field>
          <ErrorMessage component="p" name="isActive" className="red" />
        </div>
        <Link to="/items" className="ui button">
          Back to Items List
      </Link>
        <button className="ui button primary" type="submit">
          {this.state.id !== '' ? 'Update' : 'Submit'}
        </button>
      </Form>
    );
  }

  render() {
    return (
      <Grid>
        <Loader isLoading={this.props.item.isLoading} />
        <Grid.Row centered>
          <Grid.Column width="8">
            <h2 data-test="page-heading">
              {this.state.id !== '' ? 'Update' : 'Add'}
              {' '}
              Item
            </h2>
            <Formik
              enableReinitialize
              initialValues={this.state.initialValues}
              validationSchema={this.formValidation()}
              onSubmit={(values) => {
                this.formSubmitHandler(values);
              }}
            >
              {this.renderForm}
            </Formik>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }
}

function mapStateToProp({ item }) {
  return {
    item,
  };
}

export default connect(mapStateToProp, { getItem, updateItem, addItem })(UnconnectedAddItem);
