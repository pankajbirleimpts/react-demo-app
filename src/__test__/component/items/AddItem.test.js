import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { BrowserRouter } from 'react-router-dom';
import AddItem, { UnconnectedAddItem } from '../../../components/items/AddItem';
import { storeFactory, findByAttr } from '../../test/TestUtils';



const setup = (props) => {
    const wrapper = shallow(
        <UnconnectedAddItem {...props} />,
    )
    return wrapper;
};

describe('Add item Flow', () => {
    const props = {
        item: {
            isLoading: false,
        },
        history: {
            replace: jest.fn(),
        },
        getItem: jest.fn(),
        updateItem: jest.fn(),
        addItem: jest.fn(),
        match: {
            params: {
                id: undefined,
            }
        }
    }
    test('Snapshot of AddItem component', () => {
        const wrapper = setup(props);
        expect(wrapper).toMatchSnapshot();
    });


    test('Check the isLoading in props', () => {
        const wrapper = setup(props);
        const componentProps = wrapper.instance().props;
        expect(componentProps.item.isLoading).toBe(props.item.isLoading);
    });

    test('Check getItem action should not be called props', () => {
        const wrapper = setup(props);
        wrapper.instance().componentDidMount();
        expect(props.getItem).not.toHaveBeenCalled();
    });

    test('Check page titile should be `Add item`', () => {
        const wrapper = setup(props);
        const pageTitle = findByAttr(wrapper, 'page-heading');
        expect(pageTitle.text()).toEqual('Add Item');
    });

    test('Checking the form values ', async () => {
        const formValues = {
            itemName: 'Kaju Kadi',
            description: 'This is Kaju kadi',
            category: "Lunch",
            amount: 30,
            isActive: 'Active',
        };

        const wrapper = mount(<BrowserRouter>
            <UnconnectedAddItem {...props} />
        </BrowserRouter>);

        const itemNameInput = findByAttr(wrapper, 'itemName-input').find('input').first();
        const descriptionInput = findByAttr(wrapper, 'description-input').find('textarea').first();
        const amountInput = findByAttr(wrapper, 'amount-input').find('input').first();
        const categoryInput = findByAttr(wrapper, 'category-input').find('select').first();
        const isActiveInput = findByAttr(wrapper, 'isActive-input').find('select').first();

        await act(async () => {
            itemNameInput.simulate('change', { persist: () => { }, target: { name: 'itemName', value: formValues.itemName } });
        });

        await act(async () => {
            descriptionInput.simulate('change', { persist: () => { }, target: { name: 'description', value: formValues.description } });
        });

        await act(async () => {
            amountInput.simulate('change', { persist: () => { }, target: { name: 'amount', value: formValues.amount } });
        });

        await act(async () => {
            categoryInput.simulate('change', { persist: () => { }, target: { name: 'category', value: formValues.category } });
        });

        await act(async () => {
            isActiveInput.simulate('change', { persist: () => { }, target: { name: 'isActive', value: formValues.isActive } });
        });

        await act(async () => {
            wrapper.find('form').simulate('submit');
        });
        expect(props.addItem).toHaveBeenCalledWith(formValues, expect.anything());
    });
});

describe('Update item Flow', () => {
    const props = {
        item: {
            isLoading: false,
        },
        history: {
            replace: jest.fn(),
        },
        getItem: jest.fn(),
        updateItem: jest.fn(),
        addItem: jest.fn(),
        match: {
            params: {
                id: 101,
            }
        }
    }

    test('Check the isLoading in props', () => {
        const wrapper = setup(props);
        const componentProps = wrapper.instance().props;
        expect(componentProps.item.isLoading).toBe(props.item.isLoading);
    });

    test('Check getItem action should be called props', () => {
        const wrapper = setup(props);
        wrapper.instance().componentDidMount();
        expect(props.getItem).toHaveBeenCalled();
    });

    test('Check page titile should be `Update item`', () => {
        const wrapper = setup(props);
        wrapper.instance().setState({ id: props.match.params.id });
        const pageTitle = findByAttr(wrapper, 'page-heading');
        expect(pageTitle.text()).toEqual('Update Item');
    });

    test('Checking the form values ', async () => {

        const formValues = {
            itemName: 'Kaju Kadi',
            description: 'This is Kaju kadi',
            category: "Lunch",
            amount: 30,
            isActive: 'Active',
        };

        const wrapper = mount(<BrowserRouter>
            <UnconnectedAddItem {...props} />
        </BrowserRouter>);
        wrapper.instance().setState({
            id: props.match.params.id,
            initialValues: formValues,
        });

        const itemNameInput = findByAttr(wrapper, 'itemName-input').find('input').first();
        const descriptionInput = findByAttr(wrapper, 'description-input').find('textarea').first();
        const amountInput = findByAttr(wrapper, 'amount-input').find('input').first();
        const categoryInput = findByAttr(wrapper, 'category-input').find('select').first();
        const isActiveInput = findByAttr(wrapper, 'isActive-input').find('select').first();

        await act(async () => {
            itemNameInput.simulate('change', { persist: () => { }, target: { name: 'itemName', value: formValues.itemName } });
        });

        await act(async () => {
            descriptionInput.simulate('change', { persist: () => { }, target: { name: 'description', value: formValues.description } });
        });

        await act(async () => {
            amountInput.simulate('change', { persist: () => { }, target: { name: 'amount', value: formValues.amount } });
        });

        await act(async () => {
            categoryInput.simulate('change', { persist: () => { }, target: { name: 'category', value: formValues.category } });
        });

        await act(async () => {
            isActiveInput.simulate('change', { persist: () => { }, target: { name: 'isActive', value: formValues.isActive } });
        });

        await act(async () => {
            wrapper.find('form').simulate('submit');
        });
        expect(props.updateItem).toHaveBeenCalledWith(props.match.params.id, formValues, expect.anything(), true);
    });
}); 