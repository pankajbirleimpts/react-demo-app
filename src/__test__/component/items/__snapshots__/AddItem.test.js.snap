// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add item Flow Snapshot of AddItem component 1`] = `
<Grid>
  <Loading
    isLoading={false}
  />
  <GridRow
    centered={true}
  >
    <GridColumn
      width="8"
    >
      <h2
        data-test="page-heading"
      >
        Add
         
        Item
      </h2>
      <Formik
        enableReinitialize={true}
        initialValues={
          Object {
            "amount": "",
            "category": "",
            "description": "",
            "isActive": "",
            "itemName": "",
          }
        }
        onSubmit={[Function]}
        validationSchema={
          ObjectSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_defaultDefault": [Function],
            "_deps": Array [],
            "_excludedEdges": Array [],
            "_exclusive": Object {},
            "_mutate": undefined,
            "_nodes": Array [
              "isActive",
              "amount",
              "category",
              "description",
              "itemName",
            ],
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "object",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "fields": Object {
              "amount": NumberSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "number",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "category": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "description": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "max": true,
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "isActive": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "itemName": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "max": true,
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
            },
            "tests": Array [],
            "transforms": Array [
              [Function],
            ],
          }
        }
      >
        <Component />
      </Formik>
    </GridColumn>
  </GridRow>
</Grid>
`;
